### PR TITLE
[e2e] Fixed bug in e2e for ios where it doesn't exit.

### PR DESCRIPTION
**Pull request by gaaclarke**
## Description

Fixed problem with XCTest in XCode 11.3 where the testing bundles were getting opened multiple times which interfered with the singleton logic for E2EPlugin.

## Related Issues

- https://github.com/flutter/flutter/issues/48601

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by digiter** [comment]
Defer to @collinjackson as I'm not familiar with objc.

**Comment by jmagman** [comment]
> Fixed problem with XCTest in XCode 11.3 where the testing bundles were getting opened multiple times which interfered with the singleton logic for E2EPlugin.

You mean the bundles were being loaded multiple times?  That seems odd.  Why did that prevent exit?

**Comment by gaaclarke** [comment]
> You mean the bundles were being loaded multiple times? That seems odd. Why did that prevent exit?

This will loop forever since the test results are set on the other instance of the plugin.

https://github.com/flutter/plugins/blob/1d16548e29d85297cbd7830d8ae5f57f7f068062/packages/e2e/ios/Classes/E2EIosTest.m#L8

**Comment by jmagman** [comment]
> the testing bundles were getting opened multiple times which interfered with the singleton logic for E2EPlugin.

Ohhh because e2e is a static library and the dispatch_once_t is being set in both the Runner and RunnerTest bundles.

File: packages/e2e/ios/Classes/E2EPlugin.m [hunk 1]
```diff
@@ -27,9 +27,14 @@ - (instancetype)initForRegistration {
 }
 
 + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+}
+
+- (void)setupChannels:(id<FlutterBinaryMessenger>)binaryMessenger {
```
Comment (jmagman) [inline packages/e2e/ios/Classes/E2EPlugin.m]: Maybe add a comment here for why you did this, mentioning the static issue if e2e plugin is linked as a static library.

**Comment by gaaclarke** [comment]
I've let this go dormant, sorry.  It appears that this only fixes the hang for the case where tests are passing.  It still hangs if one of the tests hang.  I'm going to update the issue with better reproduction.

File: packages/e2e/ios/Classes/E2EPlugin.m (hunk 1 above)
Comment (gaaclarke) [reply packages/e2e/ios/Classes/E2EPlugin.m]: Done.

**Pull request merged**